fastjsonschema/generator.py

```diff
@@ -187,14 +212,111 @@ def generate_ref(self):
+        :param source_line: python-line source
+        :param deduplicate: check for repetitive code
+        :return:
+        """
+        if deduplicate:
+            for ind, code in reversed(self._code):
+                if ind is None:
+                    break
+                if ind == indentation:
+                    commented = '# {}'.format(source_line)
+                    if code in (source_line, commented):
+                        self._code.append((indentation, commented))
+                        return
+                    break
+        self._code.append((indentation, source_line))
+
+    def code_break(self):
+        self._emit(None, None)
+
+    def mark_type(self, variable, types):
+        for ind, code in reversed(self._code):
+            if ind is None:
+                break
+            if ind is MARKER:
+                code['types'][variable] = types
+                return
+        self._emit(MARKER, defaultdict(types={variable: types}))
+
+    def get_mark(self, mark):
+        for ind, code in reversed(self._code):
+            if ind is None:
+                return None
+            if ind is MARKER:
+                return code[mark]
+        return None
+
+    def has_type(self, atype):
+        types = self.get_mark('types')
+        return types and self._variable in types and atype in types[self._variable]
+
+    @contextmanager
+    def in_context(self, path):
+        """
+        Tracks the context path of the schema definition.
+        :param path: Element to add to the path upon invocation
+        :return:
+        """
+        self._context.append(path)
+        try:
+            yield
+        finally:
+            self._context.pop()
+
+    def context_path(self):
+        path = self._context.copy()
+        return '{} in schema{}'.format(repr(path.pop()), ''.join('[{}]'.format(repr(item)) for item in path))
+
+    @contextmanager
+    def probing(self):
+        """
+        Context manager to keep track of probing blocks to support quick exception generation.
 
+        When using exceptions in constructs like `oneOf` parts of input are probed if they satisfy some definition.
+        One can spot this be `self.l('except JsonSchemaException: pass')` catching exceptions with `pass` statement.
+        In those cases the message held by exception is meaningless so expanding them is useless. That block of code
+        should be executed in the context of:
+
+        ... code-block:: python
+            with self.probing():
+                for definition_item in self._definition['oneOf']:
+                    ..
+        """
+        self._probing += 1
+        try:
+            yield
+        finally:
+            self._probing -= 1
+
+    @property
+    def is_probing(self):
+        return bool(self._probing)
 
     # pylint: disable=invalid-name
     @indent
-    def l(self, line, *args, **kwds):
+    def l(self, line, *args, deduplicate=False, expand_locals=True, **kwds):  # noqa
```

Comment:
This library is using pylint.

Reply:
correct. can be removed. (as i mentioned, the library code lives in the project and i had to lint it with existing tools. when the pull request gets merged, i can happily go using the official version)

Reply:
I see. This merge request is actually a patch of changes you did for your copy of this library in your app.

Reply:
exactly